xxxx-ledger-plugin-interface-2/xxxx-ledger-plugin-interface-2.md

```diff
@@ -0,0 +1,502 @@
+
+If the `custom` data is too large, the ledger plugin MUST reject with a `MaximumCustomDataSizeExceededError`.
+
+Note that connectors MAY forward some fields of `custom` data from plugin to plugin, but generally are not expected to. All `custom` fields that were passed to `sendTransfer` MUST be passed to the transfer handler by the plugin on the receiving side. The only exception are properties which start with the underscore character (`_`), which MAY be consumed by the plugin and not passed on.
+
+###### Example
+``` js
+{
+  custom: {
+    claim: '...',
+    fulfillmentLatency: 29
+  }
+}
+```
+
+## Class: InterledgerRejectionError
+<code>class InterledgerRejectionError</code>
+
+An `InterledgerRejectionError` is a throwable object representing a rejection of an Interledger transfer. Implementations SHOULD use a class named `InterledgerRejectionError` which derives from JavaScript's built-in `Error`. However, other implementations MUST NOT rely on this and SHOULD use the `name` property to distinguish Interledger rejections from other error types.
+
+Plugins SHOULD NOT generally trigger `InterledgerRejectionError`s. Instead, the plugin SHOULD trigger a local error, such as the ones specified in this document and the hosting connector SHOULD create a suitable `InterledgerRejectionError` from the local error. This is because plugins are not necessarily aware of their ILP address and therefore may not be able to set `triggeredBy` correctly. Some plugins may have elements of a connector built-in if they are used with ledgers that don't natively support ILP. In that case, the plugin MAY trigger an `InterledgerRejectionError` since it is in effect acting as a connector.
+
+All fields described below MUST be present, however they MAY be empty.
+
+###### Fields
+| Type | Name | Description |
+|:--|:--|:--|
+| `String` | [name](#interledgerrejectionerrorname) | `'InterledgerRejectionError'` |
+| `String` | [message](#interledgerrejectionerrormessage) | Error message for local use |
+| `Object` | [ilpRejection](#interledgerrejectionerrorilprejection) | Information about the ILP rejection |
+| `String` | [ilpRejection.code](#interledgerrejectionerrorilprejection) | Machine-readable error code |
+| `String` | [ilpRejection.name](#interledgerrejectionerrorilprejection) | Human-readable description of the error code |
+| `String` | [ilpRejection.triggeredBy](#interledgerrejectionerrorilprejection) | ILP address from which the rejection originates |
+| `String[]` | [ilpRejection.forwardedBy](#interledgerrejectionerrorilprejection) | The ILP addresses of zero or more connectors who forwarded the rejection |
+| `String` | [ilpRejection.triggeredAt](#interledgerrejectionerrorilprejection) | ISO 8601 datetime. Describes when the rejection originally occurred. |
+| `Object` | [ilpRejection.additionalInfo](#interledgerrejectionerrorilprejection) | Additional details about the error |
```

Comment:
I think we need more details on what value types are allowed in here and how you're supposed to use it. Do we only support JavaScript JSON values? What's the range on supported numbers?

Also, is every implementation free to use its own key names? Can connectors modify this to add more details? What if they want to add a property but the key is already used by another application?

Overall, I'm not sure that having an object here is much more helpful than having a single `data` property that transport protocols can define the encoding for, or is otherwise assumed to be a string.

Reply:
How about adding a `data` property that is specifically for Transport Layer protocols and then specifying that `additionalInfo` can contain key-value pairs as UTF-8 strings?